include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,389 @@
+ * @author Sayan Goswami
+ *
+ * Implementation of the NSGA-II algorithm. Used for multi-objective
+ * optimization problems on arbitrary functions.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more Information.
+ */
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+#define ENSMALLEN_NSGA2_NSGA2_IMPL_HPP
+
+#include "nsga2.hpp"
+
+namespace ens {
+
+inline NSGA2::NSGA2(const size_t populationSize,
+                    const size_t maxGenerations,
+                    const double crossoverProb,
+                    const double mutationProb,
+                    const double mutationStrength,
+                    const double epsilon,
+                    const arma::vec& lowerBound,
+                    const arma::vec& upperBound) :
+    populationSize(populationSize),
+    maxGenerations(maxGenerations),
+    crossoverProb(crossoverProb),
+    mutationProb(mutationProb),
+    mutationStrength(mutationStrength),
+    epsilon(epsilon),
+    lowerBound(lowerBound),
+    upperBound(upperBound)
+{ /* Nothing to do here. */ }
+
+//! Optimize the function.
+template<typename MatType,
+         typename... ArbitraryFunctionType,
+         typename... CallbackTypes>
+std::vector<MatType> NSGA2::Optimize(std::tuple<ArbitraryFunctionType...>& objectives,
+                                     MatType& iterate,
+                                     CallbackTypes&&... callbacks)
+{
+  // Make sure for evolution to work at least four candidates are present.
+  if (populationSize < 4 && populationSize % 4 != 0)
+  {
+    throw std::logic_error("NSGA2::Optimize(): population size should be at"
+        " least 4, and, a multiple of 4!");
+  }
+
+  numObjectives = sizeof...(ArbitraryFunctionType);
+
+  // Cache calculated objectives.
+  std::vector<arma::Col<double> > calculatedObjectives;
+  // Pre-allocate space for the calculated objectives.
+  calculatedObjectives.resize(populationSize);
+
+  // Population size reserved to 2 * populationSize + 1 to accommodate
+  // for the size of intermediate candidate population.
+  std::vector<MatType> population;
+  population.reserve(2 * populationSize + 1);
+
+  // Pareto fronts, initialized during non-dominated sorting.
+  std::vector<std::vector<size_t> > fronts;
+  // Initialised in CrowdingDistanceAssignment.
+  std::vector<double> crowdingDistance;
+  // Initialised during non-dominated sorting.
+  std::vector<size_t> ranks;
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // Generate the population based on a Gaussian distribution around the given
+  // starting point.
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    population.push_back(arma::randu<MatType>(iterate.n_rows,
+        iterate.n_cols) + iterate);
```

Comment:
```suggestion
        iterate.n_cols) - 0.5 + iterate);
```

According to https://arxiv.org/pdf/1901.00577.pdf it's easy for NSGA2 to fall into a local optimum, especially if the initial population isn't good. There are various strategies we could implement to compensate for that, in my tests the fix is quite stable.